domains/pallets/executive/src/migrations.rs

```diff
@@ -0,0 +1,171 @@
+//! Runtime storage migrations for pallet-executive and system pallets.
+
+use event_segments_migration::EventSegments;
+use frame_support::traits::{Get, OnRuntimeUpgrade};
+use frame_support::weights::Weight;
+use frame_system::pallet::Config;
+use frame_system::{EventRecord, Pallet as System};
+#[cfg(not(feature = "std"))]
+use sp_std::vec::Vec;
+
+pub struct StorageCheckedMigrateToEventSegments<T>(sp_std::marker::PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for StorageCheckedMigrateToEventSegments<T> {
+    fn on_runtime_upgrade() -> Weight {
+        // The system pallet doesn't have storage versions, so we need to check if we were previously using Events storage instead.
+        // The migrations run after the upgraded System::reset_events() runs on the old storage format, which leads to an inconsistency:
+        // - the storage claims there are uncleared events, because it was set to the old event count, but
+        // - the event segments storage is empty.
+        //
+        // This migration is safe to run on any upgrade, because:
+        // - if event segments are disabled (or there are no uncleared events for any other reason), it will do nothing;
+        // - if event segments are enabled and in use, it will also do nothing.
+        if System::<T>::uncleared_event_count() > 0 && EventSegments::<T>::get(0).is_none() {
+            let weights = frame_system::migrations::migrate_from_events_to_event_segments::<T>();
```

Comment:
> We also need to add the event segment migration for the domain runtime: https://github.com/autonomys/polkadot-sdk/blob/2d0d460c37f6816d8bcdea7d6f35eeba72a42c44/substrate/frame/system/src/migrations/mod.rs#L125

@NingLin-P I'm not sure how you expected this migration to work, but without storage versions it's a bit tricky to find the right upgrade conditions. Maybe we could add these conditions to the upstream `migrate_from_events_to_event_segments` to make it easier?

Reply:
Maybe for this one, we can just the use the Hook `on_upgrade` to run the migrations without the need to do a versioned storage migration since it seems like frame_system does not use versioned storage migrations

Reply:
The storage version is not suitable in this case, because migration from `Events` to `EventSegments` doesn't change the underlying storage structure and it is controlled by the `EventSegmentSize` config item provided by the downstream without touching the upstream `frame_system`.

For this one, let's use `on_upgrade` as it is and be careful to remove the migration code after upgrade. I will add these conditions to the upstream to avoid misuse of the migration code but I would need to consider more cases.

Reply:
why cannot we write a custom migration and include it in our runtime instead? 
We can remove it after the domain runtime is upgrade

We just need to impl some trait to achive this and we dont have to update upstream if we are concerned about the misuse @NingLin-P

Reply:
nvm totally missed this migration 😆

Reply:
> For this one, let's use on_upgrade as it is and be careful to remove the migration code after upgrade. I will add these conditions to the upstream to avoid misuse of the migration code but I would need to consider more cases.

It might also be helpful to have a reverse migration from event segments to legacy events.